app/controllers/nodes_controller.rb

```diff
@@ -0,0 +1,65 @@
+class NodesController < ApplicationController
+
+	def show
+		if params[:id]
+			@node = Node.find(params[:id])
+		else
+			@node = Node.root
+		end
+
+	end
+
+	def create
+		p = create_params
+		node = Node.create!(p)
+		redirect_to node_path(node.parent_id)
+	end
+
+	def edit
+		@node = Node.find(params[:id])
+	end
+
+	def update
+		@node = Node.find(params[:id])
+		if @node.update_attributes(node_params)
+			redirect_to node_path(@node.parent_id)
+      	else 
+      		flash[:error] = "Failed to update node"
+      		redirect_to edit_node_path(@node.id)
+		end
+	end
+
+	def destroy
+		@node = Node.find(params[:id])
+
+		if @node.nodes.any?
```

Comment:
I think this validation needs to be put into the Node model, in case in the future there are other places we delete a Node e.g. through the API.

You can validate this on the association within the Node model i.e.

`has_many :nodes, class_name: 'Node', foreign_key: 'parent_id', dependent: :restrict_with_error`

The `dependent: :restrict_with_error` adds an error to the errors object if the node you are trying to delete has any associations (children).

More details: http://guides.rubyonrails.org/association_basics.html#has-many-association-reference